### PR TITLE
IBM experimental dataloaders

### PR DESCRIPTION
**Pull request by daviswer**
This PR introduces an experimental PyTorch-native dataloader from IBM that is distributed, stateful, checkpointable, composable and rescalable. It is intended for use in large-scale model pretraining, particularly in research settings where rapid iteration between datasets may be required. It automatically and invisibly handles data sharding, shuffling, subdataset weighting, checkpoint saving and loading, and more, with minimal overhead and high throughput.

- Add experimental dataset source file
- Add experimental dataloader builder, hooked into torchtitan cfg
- Update torchtitan cfg with additional dataset arg fields
- Update train script to build experimental dataloader instead of hf depending on cfg flags
- Replace the existing C4-mini example dataset with one that matches the expected formatting for the experimental dataloader
- TODO: port over unit tests as well
- TODO: preprocessing script(s) for the new dataset format
- TODO: further cleanup/iteration

### CODE REVIEW

File: torchtitan/checkpoint.py [hunk 1]
```diff
@@ -120,7 +120,6 @@ def __init__(
                 "model": ModelWrapper(model),
                 "optimizer": OptimizerWrapper(model, optimizer),
                 "lr_scheduler": lr_scheduler,
-                "dataloader": dataloader,
```
Comment (fegin) [inline torchtitan/checkpoint.py]: This will break the usage of the existing dataloader checkpoint.

**Comment by gokulavasan** [comment]
Thanks for the PR! Reviewing it now

File: torchtitan/config_manager.py [hunk 2]
```diff
@@ -329,6 +329,49 @@ def __init__(self):
...
+        self.parser.add_argument(
+            "--dataset.drop_tokens",
+            type=str,
+            default="",
+            help="Dummy token values to drop from begin/end of sequences (comma-separated ints)",
+        )
+        self.parser.add_argument(
+            "--dataset.datasets",
+            type=str,
+            default="c4_mini",
```
Comment (gokulavasan) [inline torchtitan/config_manager.py]: For multiple (sub) datasets, would this be specified as a list of str?

File: torchtitan/datasets/experimental/llama3/meta/c4_llama3_counts.csv [hunk 3]
```diff
@@ -0,0 +1,2 @@
+dataset/filename,documents,tokens
+/c4_mini/c4_mini.arrow,45000,20505558
```
Comment (gokulavasan) [inline torchtitan/datasets/experimental/llama3/meta/c4_llama3_counts.csv]: i) Are all the tokens for a single doc present in one row+col?
ii) What is the need to pretokenize the data instead of storing raw text alone?

File: torchtitan/datasets/experimental_datasets.py [hunk 4]
```diff
@@ -0,0 +1,1212 @@
...
+        states = [torch.load(os.path.join(path, x)) for x in my_fileshards]
+        self.load_state_dict(states, True)
+
+    def save_to_path(self, path: str):
+        """
+        Grab recursive shard states and save all shard states to the specified checkpoint folder
+        """
+        os.makedirs(path, exist_ok=True)
+        state = self.state_dict()
+        torch.save(state, os.path.join(path, f"loader_state_{self.rank}.pth"))
```
Comment (gokulavasan) [inline torchtitan/datasets/experimental_datasets.py]: From the description, it looks like only num_workers=1 is supported per rank. Thus self.rank would be sufficient in this case. Just to confirm, for num_workers > 1, do you foresee each worker storing their state from the worker process directly to local disk?

File: torchtitan/datasets/experimental_datasets.py [hunk 5]
```diff
@@ -0,0 +1,1212 @@
...
+        Retrieve all state and reshard flags (each worker/process saves its own state dict shard)
+        """
+        return {
+            self.statename(flag): getattr(self, flag)
+            for flag in self.state_params + self.reshard_params
+        }
+
+    def _reshard(self, sharded_list):
+        """
+        Sharded_list is a list of lists, where each "shard" sublist must have the same length.
```
Comment (gokulavasan) [inline torchtitan/datasets/experimental_datasets.py]: Can you elaborate on why "shard" sublist must have the same length? For example, when the input dataset folder is created with arrow files inside - does each file correspond to a shard? In that case, is each file expected to have the same length?

File: torchtitan/datasets/experimental_datasets.py [hunk 6]
```diff
@@ -0,0 +1,1212 @@
...
+        item_offset = shard_len * shard_offset
+        # How many list items are there in total?
+        n_items = self.load_worldsize * shard_len
+        # The indices of the flattened sharded_list that this worker owns
+        my_items = range(
+            int(n_items * self.rank / self.worldsize) - item_offset,
+            int(n_items * (self.rank + 1) / self.worldsize) - item_offset,
+        )
+        # Pull out owned items
+        return [sharded_list[i // shard_len][i % shard_len] for i in my_items]
```
Comment (gokulavasan) [inline torchtitan/datasets/experimental_datasets.py]: Does the sharded list is a basically a list of list of indices? And does this return a list of indices that is owned by this rank? If that is the case, could this be a really large list for huge datasets?

Update: Looking at the code further down, this seems to be for in-memory buffers instead of sample/doc indices?

File: torchtitan/datasets/experimental_datasets.py [hunk 7]
```diff
@@ -0,0 +1,1212 @@
...
+        if not sharded_input:
+            self.load_worldsize = len(state_dicts)
+            state_dicts = _shard_inclusive(state_dicts, self.rank, self.worldsize)
+        if self.load_worldsize == self.worldsize:
+            [
+                setattr(self, flag, state_dicts[0][self.statename(flag)])
+                for flag in self.state_params + self.reshard_params
+            ]
+        else:
+            for flag in self.reshard_params:
```
Comment (gokulavasan) [inline torchtitan/datasets/experimental_datasets.py]: For all parameters that are set as reshard params, this logic figures out which subset of indices in that list of ranges  that it owns? Apart from indices assigned to a worker, what are the other states tend to lend well with re-sharding?

And looks like state params are discarded during reshard? Would RNG state be a typical state param?

File: torchtitan/datasets/experimental_datasets.py [hunk 8]
```diff
@@ -0,0 +1,1212 @@
...
+    def state_dict(self):
+        """
+        Retrieve all state and reshard flags (each worker/process saves its own state dict shard)
+        """
+        return {
+            self.statename(flag): getattr(self, flag)
+            for flag in self.state_params + self.reshard_params
+        }
+
+    def _reshard(self, sharded_list):
```
Comment (gokulavasan) [inline torchtitan/datasets/experimental_datasets.py]: So this function doesn't place any restriction on what world_size could be in relation to load_world_size - is that correct?

File: torchtitan/datasets/experimental_datasets.py [hunk 9]
```diff
@@ -0,0 +1,1212 @@
...
+
+            # Load a point to buffer if necessary
+            if self.buffer_size < self.window_size:
+                self.buffer[self.buffer_size] = next(dataset)
+                self.buffer_size += 1
+
+            # Swap out randomly sampled value from buffer
+            i = torch.randint(self.buffer_size, (1,), generator=self.generator).item()
+            out = self.buffer[i]
+            self.buffer[i] = next(dataset)
```
Comment (gokulavasan) [inline torchtitan/datasets/experimental_datasets.py]: once the dataset runs out of data, how is the remaining items in the buffer drained?

File: torchtitan/datasets/experimental_datasets.py [hunk 10]
```diff
@@ -0,0 +1,1212 @@
...
+                    self.epochs_seen += 1
+                self.docset_index = doc_index
+                # Map doc id to shard, id in file
+                shardid, docrange, mindoc = self._get_docid(doc_index)
+
+                # Read doc
+                newpath = os.path.join(self.data, shardid)
+                path, reader = self._get_reader(path, newpath, reader)
+                # Map id in range of owned docs to new (consistently) shuffled id
+                doclcg = self._random_map_docid(docrange)
```
Comment (gokulavasan) [inline torchtitan/datasets/experimental_datasets.py]: Just to confirm, once a shard is chosen, all the docs within that shard is visited in random order but that shard is exhausted before moving to the next shard?

File: torchtitan/datasets/experimental_datasets.py [hunk 11]
```diff
@@ -0,0 +1,1212 @@
...
+        self.chunk_index = -1
+
+        # Stats
+        self.epochs_seen = -1
+        self.tokens_seen = 0
+        self.docs_seen = 0
+        self.percent_seen = 0
+        self.lcg_state = seed + rank
+
+        self.state_params = [
```
Comment (gokulavasan) [inline torchtitan/datasets/experimental_datasets.py]: I see no reshard_params, so this dataset doesn't support resharding?

**Review by gokulavasan — COMMENTED** [review]
Finished one pass through the PR. Left comments to understand the code better

**Review by gokulavasan — COMMENTED** [review]
Finished one pass through the PR. Left comments to understand the code better

**Review by gokulavasan — COMMENTED** [review]
Finished one pass through the PR. Left comments to understand the code better

File: torchtitan/config_manager.py (hunk 2 above)
Comment (daviswer) [reply torchtitan/config_manager.py]: For multiple subdatasets, we have it set up to take a string of comma-separated names. There's an example comment in `train_configs/llama3_8b.toml`:
```
datasets = "CommonCrawl,C4,Github,Wikipedia,Books,ArXiv,StackExchange"
dataset_weights = "67,15,4.5,4.5,4.5,2.5,2"
```
Of course this can be changed if another format is more convenient

File: torchtitan/datasets/experimental/llama3/meta/c4_llama3_counts.csv (hunk 3 above)
Comment (daviswer) [reply torchtitan/datasets/experimental/llama3/meta/c4_llama3_counts.csv]: 1. The expected format is for each document to occupy one row in the pyarrow file. Because docs are different lengths, there's not really any consistent notion of columns anymore (or viewed another way, the doc rows all occupy the single "tokens" column of the shard file)
2. Right now we pretokenize to enable chunking - currently, we can yield portions of a document at a time, without ever having to load the whole doc into memory at once, which is helpful in the case of extremely long docs. This gets more difficult for raw text since we have to split on tokens rather than on characters, but we don't know the token boundaries until after tokenization. Support for dynamic tokenization is on the todo list though!

File: torchtitan/datasets/experimental_datasets.py (hunk 4 above)
Comment (daviswer) [reply torchtitan/datasets/experimental_datasets.py]: Yes that's the eventual plan for supporting num_workers > 1

File: torchtitan/datasets/experimental_datasets.py (hunk 8 above)
Comment (daviswer) [reply torchtitan/datasets/experimental_datasets.py]: That's correct

File: torchtitan/datasets/experimental_datasets.py (hunk 5 above)
Comment (daviswer) [reply torchtitan/datasets/experimental_datasets.py]: I suppose it's not strictly necessary that shards have the same length, we just take that shortcut because this function is only ever used to reassign per-worker data structures from `load_worldsize` to `worldsize`. So far we've been able to assume that workers all hold the same fundamental data structures so this hasn't been an issue. This doesn't get called on any actual data shard files - apologies for unclear naming/commenting here

File: torchtitan/datasets/experimental_datasets.py (hunk 6 above)
Comment (daviswer) [reply torchtitan/datasets/experimental_datasets.py]: Yes your update is correct. We never actually materialize the full list of doc indices owned by each worker, because as you say, this gets very large.

File: torchtitan/datasets/experimental_datasets.py (hunk 7 above)
Comment (daviswer) [reply torchtitan/datasets/experimental_datasets.py]: Yes RNG state is a typical state param. Reshard params are generally buffers or sets of stateful subworkers

File: torchtitan/datasets/experimental_datasets.py (hunk 9 above)
Comment (daviswer) [reply torchtitan/datasets/experimental_datasets.py]: The expectation is a sublayer that never runs out - currently, we just loop over epochs forever

File: torchtitan/datasets/experimental_datasets.py (hunk 10 above)
Comment (daviswer) [reply torchtitan/datasets/experimental_datasets.py]: Technically, once a shard is chosen, all the docs within the _fraction of the shard owned by this worker_ are visited. But yes that's broadly correct.

File: torchtitan/datasets/experimental_datasets.py (hunk 11 above)
Comment (daviswer) [reply torchtitan/datasets/experimental_datasets.py]: Correct, this one particular dataset needs to sit under a `Scalable_Shard_Dataset` to support resharding

**Comment by daviswer** [comment]
Thanks for taking a look! Left a bunch of responses, hopefully this brings clarity

File: torchtitan/datasets/experimental_datasets.py [hunk 12]
```diff
@@ -0,0 +1,1212 @@
...
+    """
+    In cases where len(itemlist) % worldsize != 0, allow for fractional ownership of items,
+    and return the span including all owned items, fractional or otherwise.
+    """
+    start = math.floor(len(itemlist) * rank / worldsize)
+    end = math.ceil(len(itemlist) * (rank + 1) / worldsize)
+    return itemlist[start:end]
+
+
+class _Stateful_Dataset(data.IterableDataset):
```
Comment (andrewkho) [inline torchtitan/datasets/experimental_datasets.py]: nit: can we stick with Camel case and remove underscores between words?

**Comment by daviswer** [comment]
Updating the dataloader to match the latest version in our public repo. Re-syncs to latest main, and incorporates many previously discussed features:

- Remove dependence on the separate metadata "count file" in the dataset directory
- Support n_workers > 1. This is accomplished by shunting all path/rank-dependent setup out of initialization and into a new setup() method, which runs after init but before any other op
- Support HF-style parquet raw text datasets, with tokenization on the fly (for reasonably sized documents/shardfiles)
- Support non-flat data directories: all legal files under the specified location will be included, regardless of depth or location. Enables simple weight-free dataset mixing via a single `StreamingDocDataset` on the parent directory
- `SamplingDataset` and `ScalableShardDataset`are now implemented as proper `_WrapperDatasets`, reflecting the intended modular usage
- Fix Weird_Separated_Camel_Case naming convention in favor of ProperClassNaming
- Allow `PreloadBufferDataset` to shrink back down to the desired size after rescaling to a smaller number of workers
